src/dune_pkg/fetch.ml

```diff
@@ -56,16 +56,59 @@ module Curl = struct
            ([ Pp.textf "curl returned an invalid error code %d" exit_code ] @ stderr)))
     else (
       Path.unlink_no_err stderr;
-      match Int.of_string http_code with
+      (* Since we had to quote the return http_code, we need to strip it *)
+      match
+        let open Option.O in
+        String.drop_prefix_and_suffix ~prefix:"\"" ~suffix:"\"" http_code
+        >>= Int.of_string
+      with
       | None ->
         Error
           (User_message.make
              [ Pp.textf "curl returned an HTTP code we don't understand: %S" http_code ])
+      | Some 200 -> Ok ()
       | Some http_code ->
-        if http_code = 200
-        then Ok ()
-        else
-          Error (User_message.make [ Pp.textf "download failed with code %d" http_code ]))
+        Error (User_message.make [ Pp.textf "download failed with code %d" http_code ]))
+  ;;
+end
+
+module Tar = struct
+  let bin = lazy (Bin.which ~path:(Env_path.path Env.initial) "tar")
+
+  let run ~temp_dir input output =
+    let bin =
+      match Lazy.force bin with
+      | Some p -> p
+      | None -> User_error.raise [ Pp.text "tar not available in PATH" ]
+    in
+    let args =
+      [ "-xf"; Path.to_string input; "--force-local"; "-C"; Path.to_string output ]
```

Comment:
I think this only works if the copy of `tar` is smart enough to recognize the archive format automatically (GNU tar does, not sure about other implementations), it would probably be safer to specify the appropriate archive option or use the unarchiving tool directly.

Reply:
I checked freebsd tar and that appears to also determine the archive format automatically, from what I read in the manual. (At least as long as it is .gz I suppose). This also is the case on windows from testing.

Reply:
Can you also check `.bz2` on BSDs? topkg and dune-release create these (not that I am particularly fond of bzip2 but there is a significant amount of bzipped tarballs).

Reply:
For now, I will leave it as is. I can make a follow up PR where we can do the detection of the archive format by inspecting the header bytes. This is probably better than relying on file extensions. However it would bloat this PR too much to include it here.